Conversionlint

spzala opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/kubernetes/blob/master/CONTRIBUTING.md and developer guide https://github.com/kubernetes/kubernetes/blob/master/docs/devel/development.md
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
This PR adds a comment placeholder for the exported or public functions in conversion generator files. Such functions without a comment results into golint failure in various generated files. The changes in this patch takes care of about 36 related lint failures. 
Given below is an example lint error, 
zz_generated.conversion.go:91:1: exported function Convert_v1alpha1_Binding_To_servicecatalog_Binding should have comment or be unexported

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
Considering minor changes no issue is created. 

**Special notes for your reviewer**:
Hello, I am trying to fix golint failures as we see them in the kubernetes-incubator/service-catalog project. I have separate PRs opened for lint issues related to other generator code which are in separate libraries like gengo. Thanks! 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access) 
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`. 
-->
```release-note
```

spzala committed:
Fix goling failure in conversion public function doc

The exported or public functions without a comment results into golint failure
in various generated files. The changes in this patch takes care of about 36
related lint failures.
Given below is an example lint error,
zz_generated.conversion.go:91:1: exported function
Convert_v1alpha1_Binding_To_servicecatalog_Binding should have comment or be
unexported

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42781)
<!-- Reviewable:end -->

spzala commented:
cc @ k82cn

spzala commented:
cc @smarterclayton

k82cn commented:
I'd suggest not to modify generated codes: there's topic about removing generated codes from repo, and the golint should not check against generated code (run golint before generating code when build.).

duglin commented:
@k82cn the issue we run into, and the main reason behind fixing the lint issues, is that in order to have lint checkers skip generated code people need to modify their Makefiles/scripts to avoid those files/directories. This becomes an annoying addition to the build process (e.g. see all of the "grep"s in https://github.com/kubernetes-incubator/service-catalog/blob/master/Makefile#L200 ).  We'd like to avoid this extra noise/annoyance.

duglin commented:
as to running it before we build... that assume someone always goes a "make clean" type of operation before they build, which is not true for many devs. I suspect many will just do "make" and assume the Makefile dependencies will only recompile/rebuild the minimum - which in many cases means the generated files will be untouched and already there.

k82cn commented:
Can we try this https://github.com/golang/lint/pull/148 ?

duglin commented:
That sounds really neat, but isn't that just side-stepping the issue? Sure we can avoid the Makefile tweaks and avoid the lint errors, but I'm not sure why (if someone is willing to do the work) we don't want to generate lint-approved code?

caesarxuchao commented:
I think this is a pure improvement, no matter if in the future we will stop checking in generated code or if we will run golint at a different stage.

@wojtek-t I believe you are an approver for the generators. Could you give it a pass? Thanks.

k82cn commented:
> I'm not sure why (if someone is willing to do the work) we don't want to generate lint-approved code?

Personally, code style improvement is always welcome, I used to create several PRs :). For generated codes, suppose contributor/developer will not go through the detail; so I think it's OK to keep it as is. Take this PR as example, `//Convert_v1beta1_Cluster_To_federation_Cluster ...` seems not helpful to other contributor except making golint happy; so `//golint:ignore` maybe better if it's included by golint which k8s's using.

k82cn commented:
@k8s-bot ok to test

spzala commented:
Thanks everyone for the ongoing discussion. @k82cn thanks and yes, as you noticed and I also mentioned in the very beginning of the PR commit message, the comments are place holder for now. If we find a better doc to put, we sure can do it now or in future but for now at least it will take care of lint failures. Thanks!!

spzala commented:
@k8s-bot verify test this

k82cn commented:
that's great. I'm ok with that:)

> 在 2017年3月10日，07:01，Sahdev Zala <notifications@github.com> 写道：
> 
> Thanks everyone for the ongoing discussion. @k82cn thanks and yes, as you noticed and I also mentioned in the very beginning of the PR commit message, the comments are place holder for now. If we find a better doc to put, we sure can do it now or in future but for now at least it will take care of lint failures. Thanks!!
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

spzala commented:
@k82cn :) thanks much!!

spzala commented:
I am not sure why Jenkins verification fails, log didn't help much :( I will try to rerun command. If anyone has idea, please help. Thanks!

spzala commented:
@k8s-bot verify test this

k82cn commented:
@spzala , please run `./hack/update-staging-client-go.sh`.

spzala commented:
@k82cn sure, thank you much!! I did run it after getting required deps and that created bunch of generated files changes, updating patch accordingly. Let's see how it goes.

k82cn commented:
@spzala , can you help to merge into only 3 commit?

* The diff for `cmd/libs/go2idl/conversion-gen/generators/conversion.go`
* The diff for generated codes
* The diff for `./hack/update-staging-client-go.sh`

spzala commented:
@k8s-bot bazel test this

spzala commented:
@k8s-bot cvm gce e2e test this

spzala commented:
@k82cn sure, I have reverted my last commit and checked-in changes for generated files and  /hack/update-staging-client-go.sh as separate patches..and changes for cmd/libs/go2idl/conversion-gen/generators/conversion.go was already on it's own commit. Hope it looks good now. Please let me know if I didn't do it right. Thanks!!

k82cn commented:
@spzala , still 7 commits :). Usually, I run command `git reset git_hash_id_xxx` to rollback and `git push --force` to over-write old commits.

k82cn commented:
@spzala , I'm also an IBMer in China, you can ST me :).

spzala commented:
@k82cn :-) thanks! I was afraid of using force, but it looks much cleaner now. As you can notice there are only two commits. After reset to first (real code change) commit, I ran make clean and make/make test, and submitted second commit of generated code changes. But afterwards ./hack/update-staging-client-go.sh didn't produce any diff (my second commit seems has diff for all generate code). Does that sounds OK? Thanks!

k82cn commented:
The files in `staging` were updated; that's ok to me :).

k82cn reviewed (approved):
LGTM :).

spzala commented:
@k82cn :) It was lots of good learning for me in this PR. Thanks much!!

k82cn commented:
@spzala , my pleasure.

spzala commented:
Hello @wojtek-t  This PR is around for a while and has received positive reviews.  So just a gentle reminder on review/merge, considering KubeCon next week I guess it will be busy week and delay on this PR. /cc @caesarxuchao  @k82cn Thanks!!

spzala commented:
@k82cn Hi Klaus, since this PR is in a long review cycle, is there any good way I can get someone's attention to it, like do I need to contact any SIG or be in SIG meeting and request for review..? Thanks!!

caesarxuchao commented:
@wojtek-t could you take a look at this PR? It's making the generated conversion code passing golint. @spzala you mentioned it's fixing the issues in kubernetes-incubator/service-catalog project, could you link to the issue? Thanks.

spzala commented:
@caesarxuchao thank you so much!! @wojtek-t @k82cn sure, the s-c issue link is https://github.com/kubernetes-incubator/service-catalog/issues/346#issuecomment-292281473 It requires multiple changes and this PR fixes on of the issues. Also, just fyi, I have also opened PRs in other libraries like gengo, go (kubernetes/gengo#40 kubernetes/gengo#41 ugorji/go#190). Thanks!!

wojtek-t commented on cmd/libs/go2idl/conversion-gen/generators/conversion.go:
```diff
@@ -578,6 +578,7 @@ func (g *genConversion) generateConversion(inType, outType *types.Type, sw *gene
 		}
 	} else {
 		// Emit a public conversion function.
+		sw.Do("//"+nameTmpl+" ... \n", args)
```
Can you add space after "//"? I mean:
```sw.Do("// "+nameTmpl+" ... \n", args) ```

Also, maybe change to slightly better comment (for doc purpose), e.g.:
s/... /is autogenerated conversion function/
or sth like that?

wojtek-t commented:
@caesarxuchao @spzala - sorry for delay, I appearently missed this PR.

I added some minor comment, but in general I'm in favor of this PR.

spzala commented:
@wojtek-t  No problem at all and your suggestions sounds good. I will be updating PR with changes soon. Thanks!!

spzala commented:
WIP

spzala commented on api/openapi-spec/swagger.json:
```diff
@@ -2,7 +2,7 @@
   "info": {
    "title": "Kubernetes",
-   "version": "v1.7.0"
+   "version": "v1.6.0"
```
Hello, I am not sure of this change and it looks bit strange to me. I didn't make any manual change, while running 'git commit', kube gave a message that "ERROR!
Openapi spec needs to be updated.
To regenerate the spec, run:
  hack/update-openapi-spec.sh
"  and I guess this change in version is probably an output of "hack/update-openapi-spec.sh"

spzala replied:
I couldn't commit any changes without this change. This is how it goes:
------------------------------------------------
I run ./hack/update-staging-client-go.sh
which gives me:
“Commit your changes in another terminal and then continue here by pressing enter.
“ 
I run “git commit -a” and that gives me:
ERROR!
Openapi spec needs to be updated.
To regenerate the spec, run:
  hack/update-openapi-spec.sh

ERROR!
Federation OpenAPI spec needs to be updated.
To regenerate the spec, run:
  hack/update-federation-openapi-spec.sh

Aborting commit
--------------------------------------------
I am trying to cheat if can 'commit' without the swagger version change, let's see if it works. Thanks!!

caesarxuchao replied:
if you commit all your changes, and then run hack/update-all.sh, does that work?

I disabled all the git pre-commit hooks in my environment. Not sure if that's a good style though :)

spzala replied:
@caesarxuchao Thanks a lot for the tip :) !! So no with pre-commit hook it forces me to make the openapi changes before I make any commit on generated code changes. But now I am actually going to try without hook, let's see how that goes :). Agree may not be a good style, but honestly with hooks it's taking so much time specially when we have multiple commits in a single PR.

spzala commented on federation/apis/openapi-spec/swagger.json:
```diff
@@ -2,7 +2,7 @@
   "info": {
    "title": "Generic API Server",
-   "version": "v1.7.0"
+   "version": "v1.6.0"
```
Same here - 
I am not sure of this change and it looks bit strange to me. I didn't make any manual change, while running "git commit" kube gave an error message:
"ERROR!
Federation OpenAPI spec needs to be updated.
To regenerate the spec, run:
  hack/update-federation-openapi-spec.sh" and I guess this change in version is probably an output of "hack/update-federation-openapi-spec.sh"  Thanks!

wojtek-t commented on cmd/libs/go2idl/conversion-gen/generators/conversion.go:
```diff
@@ -580,6 +580,7 @@ func (g *genConversion) generateConversion(inType, outType *types.Type, sw *gene
 		}
 	} else {
 		// Emit a public conversion function.
+		sw.Do("// "+nameTmpl+" is an autogenerated conversion function. \n", args)
```
Please remove space at the end of line, i.e.
s/function. \n/function.\n"/

spzala replied:
@wojtek-t  :( ohhh sorry didn't notice it.

spzala replied:
And sounds great, I will revert those version related changes and as you said that should make Jenkins happy. Thanks!!

wojtek-t replied:
Seems still not done.

wojtek-t reviewed (commented):
Your problems seem to be related to those "v1.6.0" changes - you can try manually reverting those and it should fix the problem (I don't know why this happened).

spzala committed:
Conversion generated code changes for golint fix

The exported or public functions requires a doc comment to pass golint.
This commit has changes of conversion generated code. The actual doc
changes are added into a separate commit for a clean review.

spzala commented:
@caesarxuchao @wojtek-t @k82cn so without having pre-commit hooks I could commit without openaip related unnecessary changes. Hopefully all verification will pass. Thanks!!

k82cn commented:
@spzala , thanks for your PR :). LGTM.

spzala commented:
@k82cn :) no problem, Thank You for all your help and reviews!!

wojtek-t commented:
/lgtm
/approve
